### PR TITLE
1799 adding resume arg

### PR DESCRIPTION
**Pull request by melias**

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/BenchmarkDotNet?pullRequest=2164) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/BenchmarkDotNet?pullRequest=2164) before we can accept your contribution.<br/><br/>:x: melias [sign now](https://cla.dotnetfoundation.org/dotnet/BenchmarkDotNet?pullRequest=2164)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/BenchmarkDotNet?pullRequest=2164) it.</sub>

File: src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs [hunk 1]
```diff
@@ -84,6 +85,26 @@ internal static Summary[] Run(BenchmarkRunInfo[] benchmarkRunInfos)
 
                     foreach (var benchmarkRunInfo in supportedBenchmarks) // we run them in the old order now using the new build artifacts
                     {
+                        if (benchmarkRunInfo.Config.Options.IsSet(ConfigOptions.Resume))
+                        {
+                            var directoryInfo = new DirectoryInfo(rootArtifactsFolderPath);
+                            var lastFiles = directoryInfo.GetFiles($"{title.Split('-')[0]}*");
+                            if (lastFiles.Count() > 1)
+                            {
+                                var lastUpdatedFile = lastFiles[lastFiles.Length - 2];
```
Comment (timcassell) [inline src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs:94]: You need to sort by date for this to work. `.OrderBy(p => p.CreationTime)`

File: src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs [hunk 2]
```diff
@@ -84,6 +85,26 @@ internal static Summary[] Run(BenchmarkRunInfo[] benchmarkRunInfos)
...
+                                var text = File.ReadAllText(lastUpdatedFile.FullName);
+                                var regex = new Regex("--benchmarkId (.*?) in");
+                                var match = regex.Match(text);
+                                if (match.Success)
+                                {
+                                    var parsedId = int.Parse(match.Groups[1].Value);
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs:100]: the discovery of `parseId` should be moved out of the loop, otherwise we are going to repeat it for all benchmarks and we want to do it only once

**Review by adamsitnik — COMMENTED** [review]
@melias thank you for your contribution! PTAL at my comments.

File: src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs (hunk 2 above)
Comment (melias) [reply src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs:100]: > the discovery of `parseId` should be moved out of the loop, otherwise we are going to repeat it for all benchmarks and we want to do it only once

In the loop there is `benchmarkRunInfo` and I can know if the option `resume` is set. 
If I follow your suggestion, Do you have any idea how can I do this? (know if set resume out of loop)

File: src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs (hunk 1 above)
Comment (melias) [reply src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs]: @timcassell thanks. I already committed your suggestion.

File: src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs (hunk 2 above)
Comment (adamsitnik) [reply src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs:101]: > Do you have any idea how can I do this? (know if set resume out of loop)

Before the loop starts, you can check if any of the benchmarks present in `supportedBenchmarks` has this flag set to true.

Something like this:


```cs
if (supportedBenchmarks.Any(benchmark => benchmark.Config.Options.IsSet(ConfigOptions.Resume))
```

I know it's not obvious that we don't check it for all of the benchmarks, but this is something that we frequently do in BDN.

File: src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs [hunk 3]
```diff
@@ -84,6 +85,27 @@ internal static Summary[] Run(BenchmarkRunInfo[] benchmarkRunInfos)
...
                     foreach (var benchmarkRunInfo in supportedBenchmarks) // we run them in the old order now using the new build artifacts
                     {
+                        if (benchmarkRunInfo.Config.Options.IsSet(ConfigOptions.Resume))
+                        {
+                            var directoryInfo = new DirectoryInfo(rootArtifactsFolderPath);
+                            var lastFiles = directoryInfo.GetFiles($"{title.Split('-')[0]}*");
+                            if (lastFiles.Count() > 1)
+                            {
+                                var lastFilesSorted = lastFiles.OrderBy(o => o.LastWriteTime).ToArray();
+                                var lastUpdatedFile = lastFilesSorted[lastFiles.Length - 2];
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs:95]: `lastFiles.Length - 2` is wrong, as it's not last, but one before last.

```suggestion
                                var lastUpdatedFile = lastFiles.OrderByDescending(file => file.LastWriteTime).First();
```

Comment (melias) [reply src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs]: > `lastFiles.Length - 2` is wrong, as it's not last, but one before last.

I know that. The **last is the current file** and we need one before the last where we can get the last benchmarkId.

Comment (melias) [reply src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs]: Sure @adamsitnik I'll do it now, thanks.

File: tests/BenchmarkDotNet.IntegrationTests/BenchmarkSwitcherTest.cs [hunk 4]
```diff
@@ -332,6 +332,21 @@ public void WhenUserCreatesStaticBenchmarkMethodWeDisplayAnError_FromAssembly()
...
+
+            var types = new[] { typeof(WithDryAttributeAndCategory) };
+            var switcher = new BenchmarkSwitcher(types);
+
+            var results = switcher.Run(new[] { "--resume", "--filter", "*WithDryAttribute*" }, config);
+
+            Assert.Single(results);
+            Assert.Single(results.SelectMany(r => r.BenchmarksCases));
+        }
+
```
[the remark this answers is not in the document]
Comment (melias) [reply tests/BenchmarkDotNet.IntegrationTests/BenchmarkSwitcherTest.cs]: I modify my test at BenchmarkSwitcherTest.cs to run twice and verified that skipped the other benchmarks, but just can see it debugging I didn't find an assert for this test :(

[the remark this answers is not in the document]
Comment (adamsitnik) [reply tests/BenchmarkDotNet.IntegrationTests/BenchmarkSwitcherTest.cs]: You can implement/reuse asserts similar to 

https://github.com/dotnet/BenchmarkDotNet/blob/c02c3d82bbefa6236660a7b7405f38e739679daf/tests/BenchmarkDotNet.IntegrationTests/BenchmarkTestExecutor.cs#L67-L83

[the remark this answers is not in the document]
Comment (melias) [reply tests/BenchmarkDotNet.IntegrationTests/BenchmarkSwitcherTest.cs]: Ok @adamsitnik, Thanks again ;)

**Review by adamsitnik — APPROVED** [review]
During manual testing I've found some edge case bugs, but since we don't have much time (the hackathon ends today) I've sent some minor polishing fixes directly to your fork.

Thank you for your contribution @melias !

**Comment by melias** [comment]
> During manual testing I've found some edge case bugs, but since we don't have much time (the hackathon ends today) I've sent some minor polishing fixes directly to your fork.
> 
> Thank you for your contribution @melias !

Thanks @adamsitnik 
I'm happy to contribute and I will continue to help :)

**Comment by adamsitnik** [comment]
> Is this repository participating in hacktoberfest?

We are currently participating in https://code.devexperts.com/event/openhack22

> It looks like you need to add a label to the repository:

I don't have permissions to do that, @AndreyAkinshin could you please add the label?

@YegorStepanov in the past we had to apply Hacktoberfest labels to issues/PRs in order for the contributions to count. Please let me know if I should add these labels to any of your PRs

**Comment by AndreyAkinshin** [comment]
> I don't have permissions to do that, @AndreyAkinshin could you please add the label?

Done.

**Comment by YegorStepanov** [comment]
> YOUR PR/MRS MUST BE IN A REPO TAGGED WITH THE “HACKTOBERFEST” TOPIC, OR HAVE THE “HACKTOBERFEST-ACCEPTED” LABEL.

You can remove `Hacktoberfest` from issues, because they don't count anymore.
https://github.com/dotnet/BenchmarkDotNet/issues?q=is%3Aissue+is%3Aopen+label%3AHacktoberfest
